src/kOS/Serialization/ConfigNodeFormatter.cs

```diff
@@ -0,0 +1,184 @@
+        private static readonly ConfigNodeFormatter instance = new ConfigNodeFormatter();
+
+        public static ConfigNodeFormatter Instance
+        {
+            get 
+            {
+                return instance;
+            }
+        }
+
+        private ConfigNodeFormatter()
+        {
+        }
+
+        public ConfigNode ToConfigNode(Dump dump)
+        {
+            return ToConfigNode(ParentNode, dump);
+        }
+
+        private ConfigNode ToConfigNode(string name, List<object> list)
+        {
+            ConfigNode configNode = new ConfigNode();
+            configNode.name = name;
+
+            // ConfigNode doesn't handle arrays so we wrap everything in an extra node
+            // and give it a special name to mark this data set as an array
+            ConfigNode innerNode = new ConfigNode();
+            innerNode.name = ListKey;
+
+            var indexedList = list.Select((value, index) => new { Value = value, Index = index.ToString() });
+
+            foreach (var item in indexedList) {
+                HandleValue(innerNode, item.Index, item.Value);
+            }
+
+            configNode.AddNode(innerNode);
+
+            return configNode;
+        }
+
+        private ConfigNode ScalarToConfigNode(string name, string type, string value)
+        {
+            ConfigNode configNode = new ConfigNode();
+            configNode.name = name;
+
+            // ConfigNode stores all values as strings so we have to store non-string values as nodes, otherwise
+            // the type information would be lost
+            ConfigNode innerNode = new ConfigNode();
+            innerNode.name = type;
+
+            innerNode.AddValue(ValueKey, value);
+
+            configNode.AddNode(innerNode);
+
+            return configNode;
+        }
+
+        private ConfigNode ToConfigNode(string name, Dump dump)
+        {
+            ConfigNode configNode = new ConfigNode();
+
+            configNode.name = name;
+
+            foreach (var entry in dump) {
+                HandleValue(configNode, entry.Key.ToString(), entry.Value);
+            }
+
+            return configNode;
+        }
+
+        private void HandleValue(ConfigNode configNode, string key, object value)
+        {
+            if (value is Dump)
+            {
+                configNode.AddNode(key, ToConfigNode(key, value as Dump));
+            } else if (value is List<object>)
+            {
+                configNode.AddNode(key, ToConfigNode(key, value as List<object>));
+            } else if (value is int) {
+                configNode.AddNode(key, ScalarToConfigNode(key, ScalarKey, Convert.ToString(value)));
+            } else if (value is double) {
+                configNode.AddNode(key, ScalarToConfigNode(key, ScalarKey, ((double)value).ToString("R")));
+
+            } else if (value is bool) {
+                configNode.AddNode(key, ScalarToConfigNode(key, BooleanKey, Convert.ToString(value)));
+            } else
```

Comment:
(See elsewhere in the PR comments for the comment with the word "curly brace" in big letters.)

This, specifically: the case of if/else ladders, is why I'm not a fan of the formatting standard we're using on this project and actually prefer the kind you're trying to use here.  Putting all the braces on their own separate lines makes if/else ladders waste LOTS of screen height.  It makes it take 3 lines to express a single ``else if ....`` that way, as opposed to just one line the way you're doing it.

However, like it or not, that does seem to be the standard the rest of the project is using so it should probably be changed.